### DISCUSSION TITLE
Float32bit IEEE754 decoding

### DISCUSSION THREAD

**Discussion by sawacs** [arendst/Tasmota#11819] [Q&A] [answered]
Hello everyone.

At the outset, I would like to point out that I have been with Tasmota for some time with great success.

I encountered the following problem with decoding data from the LE-03MW-CT electricity meter.
This meter has a Modbus RTU in which the FLOAT data is coded in the IEEE754 standard. 
I have written a script that sends and receives data from a meter but cannot decode it. 
Converting from FLOAT32bit to DEC is fairly straightforward if you have bitwise operation at your disposal.

Does anyone have any idea to solve this problem?

General IEEE754 decoding method:
If we have FLOAT32bit, the most significant bit is number 31 and the least significant bit is number 0.
Bit 31 specifies the sign of the number: 0 is '+' and if 1 is '-'.
Bits 30 through 23 are the eight bit value for computing the mantissa.
It is counted like this: mantissa = 2 ^ (dec (bit30-bit23) -127)
Bits 22 to 0 are a fractional multiplier - always plus 1 .
It is counted in the following steps:
1+ (bit22 * 2 ^ -1 + bit21 * 2 ^ -2 + bit20 * 2 ^ -3 + ...... + bit0 * 2 ^ -23)

The total is calculated as follows: (for example 50Hz frequency)
HEX  42               48               00                00                       float32bit
BIN   0100 0010  0100 1000 0000 0000 0000 0000
IEEE754 sign: 0, mantissa: 10000100,  fraction: 100 1000 0000 0000 0000 0000

+ Sign
Mantissa 10000 100 => 132-127 = 5 => 2 ^ 5 = 32
Fraction 1 + 2 ^-1 + 2 ^-4 = 1 + 0.5 + 0.0625 = 1.5625
Result 32 * 1.6525 = + 50

**Comment by sfromis** [comment]
Standard Tasmota processing does not include bitwise operations. I see 3 options....
1) Do it backend
2) Amend Tasmota coding in C++
3) With ESP32, the new feature of the Berry scripting language include bitwise operators

**Reply by sawacs** [reply]
Ok. I have ESP32
Here is my build configuration:
===
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif
#define MAX_METERS 1
#define SML_BSIZ 64
#define USE_SML_SCRIPT_CMD
#define SML_REPLACE_VARS
#define USE_SML_M
#define SML_REPLACE_VARS
#define USE_SML_SCRIPT_CMD
#define USE_SCRIPT_WEB_DISPLAY
#define USE_SCRIPT_JSON_EXPORT
#define USE_SCRIPT_SUB_COMMAND
#define USE_SCRIPT_TASK
#define USE_SCRIPT_TIMER
#define USE_GOOGLE_CHARTS
===
What else needs to be added?
I did not find a description in the documentation on how to process data from SML to Berry scripts and how to perform bit operations, how to convert values between hex, bin and dec?

**Reply by sfromis** [reply]
`#define USE_BERRY` includes the feature.

I've got no experience with SML, and possible integration with [Berry](https://tasmota.github.io/docs/Berry-Scripting/). Berry has decimal to hex conversion, not sure about builtin hex to dec, but a function can easily be written if relevant. Obviously bitwise operations are documented in the Berry [manual](https://github.com/Skiars/berry_doc/releases/download/latest/berry_rm_en_us.pdf). Of course the point raised by barbudor of existing float support in SML being usable or not is also highly relevant.

**Comment by barbudor** [comment, marked as answer]
@sawacs I'm not sure to fully understand your need as I'm not fully specialist on Smart meter
But I can see that the decoder already supports reading floats.
So what is exactly missing or not working ?

What does the `ffffffff` specifier doesn't do for you ?

![image](https://user-images.githubusercontent.com/2996042/115560713-221c6a80-a2b5-11eb-997a-b9b7bc13a89b.png)

**Reply by sawacs** [reply]
### Thanks for showing interest!

I'm attaching my script. I ran most of the script to read only the frequency:
===
>D
m1="1,250304ffffffff@"

>B
=>sensor53 r
=>sensor53 d1

>M 1
+1,3,M,1,9600,Licznik,1,20,25030014
;,2503000e,25030010,25030012,25030130,2503013c,25030100
1,=h<hr/>
1,=h **** Licznik prądu ****
1,250304UUuuUUuu@i0:1,Częstotliwość,Hz,freq,2
;%m1%i1:1,Napięcie L1,V,L1,2
;%m1%i2:1,Napięcie L2,V,L2,2
;%m1%i3:1,Napięcie L3,V,L3,2
;%m1%i4:1,Moc T1,kW,T1,2
;%m1%i5:1,Moc T2,kW,T2,2
;%m1%i6:1,Moc całkowita,kW,KW,2
1,=h<hr/>
#
===
Of course, I read the documentation, but I believe that decoding is about converting from float to decimal...?
I compared the instructions of my LE-03MW-CT and that described in the example of SDM530 (MODBus).
In LE-03MW-CT and in SMD530, the float numbers are saved in the IEEE754 format (https://en.wikipedia.org/wiki/IEEE_754)
SMD530 (https://bg-etech.de/download/manual/SDM630Register1-5.pdf) LE-03MW-CT (LE-03MW CT full user manual) but the my script does not decode any value.
The data from the meter goes OK, because I can see them on the terminal. 
Frec. reqest 25030014 Answer  25 03 04 42 48 0a 3d cd 2e - Last 2 Bytes is CRC
After "manual" decoding on the calculator I get the correct frequency value = 50Hz

**Reply by sawacs** [reply]
-----
By the way ... If ffffffff actually decodes ieee754 then it should be added to the instruction

**Reply by sawacs** [reply]
Little update, sent script from other trials ...
This is the right one:
>D
m1="1,250304ffffffff@"

>B
=>sensor53 r
=>sensor53 d1

>M 1
+1,3,M,1,9600,Licznik,1,20,25030014
;,2503000e,25030010,25030012,25030130,2503013c,25030100
1,=h<hr/>
1,=h **** Licznik prądu ****
%m1%i0:1,Częstotliwość,Hz,freq,2
;%m1%i1:1,Napięcie L1,V,L1,2
;%m1%i2:1,Napięcie L2,V,L2,2
;%m1%i3:1,Napięcie L3,V,L3,2
;%m1%i4:1,Moc T1,kW,T1,2
;%m1%i5:1,Moc T2,kW,T2,2
;%m1%i6:1,Moc całkowita,kW,KW,2
1,=h<hr/>
#

**Reply by barbudor** [reply]
IEEE754 is THE STANDARD for floating point representation on computers

There might be some old and odd alternatives but today most machines and languages are using that because this also how FPU inside microprocessors are handling it.
But feel free to update the doc using the small pencil on the top-right.

By the way "42 48 0a 3d" is actually 50.0099983215

**Reply by sawacs** [reply]
Yee! I did it!!!
I compiled the new and fresh version 9.4.0 witch all my options and it works!
I consider the topic to be finished

Thanks a lot...
